### ISSUE TITLE
error notification

### ISSUE THREAD

**Issue by deep-42-thought** [grindsa/acme2certifier#44]
What would be the best way to get notified about unexpected errors?

Background: I once forgot to run `db_update.py` after an update and acme2certifier rightfully failed to issue certificates since then. I did not notice this, until my generic "your certificate is soon outdated" warning on the clients kicked in - which is quite some time after I broke the database. I plan to use acme2certifier in an environment, where I do not monitor all clients (because I do not run them), but still want to get notified early if something is broken.

I'm running acme2certifier as uwsgi plugin under nginx started by systemd. I'm thinking of some "emergency notification", which kicks in, when something unexpected happens. Emitting some recognizable line to the system log (which I then can filter with logwatch) is one idea I have. Would this be something, that could cover all/most cases of "unexpected errors"?

**Comment by grindsa** [comment]
Improving the error handling in acme2certifier is in my backlog. The idea is that errors (regardless of type) will be logged into system log but this is not yet always the case.  Let me think about a quick-fix that at least failed certificate enrollment will be logged  accordingly.

**Comment by grindsa** [comment]
Just to keep you updated. I played with different options and ideas. The best option for me would be to log errors with different severities into the system log:
- database errors will be logged with severity "CRITICAL"
- enrolment errors (from CA handler) will be logged with severity "ERROR"
- other errors will be logged with severity "WARNING"

The log entry starts with `acme2certifier xxx error in function`;  the error message will be attached.

I did a quick [prototype](https://github.com/grindsa/acme2certifier/tree/error_notifications) (based on current masterbranch) who implemented such kind of logging in `nonce.py` and simulated a database issue by removing write permissions from `acme_srv.db`.

```
[Fri Aug 07 08:16:11.589382 2020] [wsgi:error] [pid 17:tid 140240429516544] [remote 192.168.14.131:59004] acme2certifier database error in Nonce.generate_and_add(): attempt to write a readonly database
[Fri Aug 07 08:16:11.668357 2020] [wsgi:error] [pid 17:tid 140240311994112] [remote 192.168.14.131:59006] 192.168.14.131 /acme/newaccount {'code': 400, 'header': {}, 'data': {'status': 400, 'message': 'urn:ietf:params:acme:error:badNonce', 'detail': 'JWS has invalid anti-replay nonce: 9f0e64216fc34f70a4e24110e92aa141'}}
[Fri Aug 07 08:16:13.728923 2020] [wsgi:error] [pid 17:tid 140240429516544] [remote 192.168.14.131:59008] acme2certifier database error in Nonce.generate_and_add(): attempt to write a readonly database
[Fri Aug 07 08:16:13.817699 2020] [wsgi:error] [pid 17:tid 140240311994112] [remote 192.168.14.131:59010] 192.168.14.131 /acme/newaccount {'code': 400, 'header': {}, 'data': {'status': 400, 'message': 'urn:ietf:params:acme:error:badNonce', 'detail': 'JWS has invalid anti-replay nonce: f222e5226f4145119bbc5f9223675b35'}}
[Fri Aug 07 08:16:15.887614 2020] [wsgi:error] 
```

Feel free to play with it and let me know if this would work for your setup.

**Comment by grindsa** [comment]
implemented in [0.11](https://github.com/grindsa/acme2certifier/releases/tag/0.11)

**Comment by grindsa** [comment]
Closing issue.. Feel free to reopen in case you it is not working as expected.
